provider.go

```diff
@@ -5,110 +5,178 @@ package azuresecrets
 
 import (
 	"context"
+	"fmt"
+	"net/http"
 	"time"
 
-	"github.com/Azure/azure-sdk-for-go/services/preview/authorization/mgmt/2018-09-01-preview/authorization"
-	"github.com/Azure/go-autorest/autorest"
-	"github.com/Azure/go-autorest/autorest/azure/auth"
-	"github.com/hashicorp/vault-plugin-secrets-azure/api"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/vault/sdk/logical"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v2"
+
 	"github.com/hashicorp/vault/sdk/helper/useragent"
+
+	"github.com/hashicorp/vault-plugin-secrets-azure/api"
 )
 
-var _ api.AzureProvider = (*provider)(nil)
+// AzureProvider is an interface to access underlying Azure Client objects and supporting services.
+// Where practical the original function signature is preserved. Client provides higher
+// level operations atop AzureProvider.
+type AzureProvider interface {
+	api.ApplicationsClient
+	api.GroupsClient
+	api.ServicePrincipalClient
+
+	CreateRoleAssignment(
+		ctx context.Context,
+		scope string,
+		roleAssignmentName string,
+		parameters armauthorization.RoleAssignmentCreateParameters) (armauthorization.RoleAssignmentsClientCreateResponse, error)
+	DeleteRoleAssignmentByID(ctx context.Context, roleID string) (armauthorization.RoleAssignmentsClientDeleteByIDResponse, error)
+	ListRoleDefinitions(ctx context.Context, scope string, filter string) (result []*armauthorization.RoleDefinition, err error)
+	GetRoleDefinitionByID(ctx context.Context, roleID string) (result armauthorization.RoleDefinitionsClientGetByIDResponse, err error)
+}
+
+var _ AzureProvider = (*provider)(nil)
 
 // provider is a concrete implementation of AzureProvider. In most cases it is a simple passthrough
 // to the appropriate client object. But if the response requires processing that is more practical
 // at this layer, the response signature may different from the Azure signature.
 type provider struct {
 	settings *clientSettings
+	// HTTP Client used to make role assignment and definition requests
+	// using new arm libraries
+	httpClient *http.Client
 
 	appClient    api.ApplicationsClient
 	spClient     api.ServicePrincipalClient
 	groupsClient api.GroupsClient
-	raClient     *authorization.RoleAssignmentsClient
-	rdClient     *authorization.RoleDefinitionsClient
+	raClient     *armauthorization.RoleAssignmentsClient
+	rdClient     *armauthorization.RoleDefinitionsClient
 }
 
 // newAzureProvider creates an azureProvider, backed by Azure client objects for underlying services.
-func newAzureProvider(settings *clientSettings, passwords api.Passwords) (api.AzureProvider, error) {
+func newAzureProvider(settings *clientSettings, passwords api.Passwords) (AzureProvider, error) {
+	// @TODO see if this is still needed
```

Comment:
It looks like we're setting the user-agent request header for the role assignments/definitions clients (via the custom transport) but not for the MS graph client. If there isn't a simple way to set the user-agent for the MS graph client, then I'd say it's okay to remove this.

Fwiw, the last we talked about these user-agent headers from MSFT they weren't making use of them for attribution/analytics.

Reply:
I did some preliminary research, and there doesn't seem to be clear cut documentation and recommendations on setting the User Agent for the MSGraph Client. I am seeing some threads mention a custom header could be passed on certain requests, but since it's not the official docs I'm opting to remove setting the User Agent on MSGraph client for now since it does not look like they are making use of it.

I can circle back to this in the future if it ends up that we do prefer adding in the header and if we find official recommendations from MSFT on how to set the value!